lewis/adapters/stream.py

```diff
@@ -86,47 +79,83 @@ def handle_accept(self):
+    was possible to pass flags to ``re.compile``, this has been removed for consistency issues
+    in :class:`Var`. It is however still possible to use the exact same flags as part of the
+    regular expression. In the documentation of re_, this is outlined, simply add a group to the
+    expression that contains the flags, for example ``(?i)`` to make the expression case
+    insensitive. This special group does not count towards the matching groups used for argument
+    capture.
+
+    The optional argument_mappings can be an iterable of callables with one parameter of the
+    same length as the number of arguments of the function. The first parameter will be
+    transformed using the first function, the second using the second function and so on.
+    This can be useful to automatically transform strings provided by the adapter into a proper
+    data type such as ``int`` or ``float`` before they are passed to the function.
+
+    The return_mapping argument is similar, it should map the return value of the function
     to a string. The default map function only does that when the supplied value
-    is not None.
+    is not None. It can also be set to a numeric value or a string constant so that the
+    command always returns the same value. If it is ``None``, the return value is not
+    modified at all.
 
     Finally, documentation can be provided by passing the doc-argument. If it is omitted,
-    the docstring of the bound method is used and if that is not present, left empty.
+    the docstring of the bound function is used and if that is not present, left empty.
 
-    :param target_method: Method to be called when regex matches.
-    :param regex: Regex to match for method call.
-    :param regex_flags: Flags to pass ot re.compile, default is 0.
+    :param func: Function to be called when pattern matches or member of device/interface.
+    :param pattern: Regex to match for function call.
     :param argument_mappings: Iterable with mapping functions from string to some type.
     :param return_mapping: Mapping function for return value of method.
     :param doc: Description of the command. If not supplied, the docstring is used.
+
+    .. _re: https://docs.python.org/2/library/re.html#regular-expression-syntax
     """
 
-    def __init__(self, target_method, regex, regex_flags=0, argument_mappings=None,
-                 return_mapping=lambda x: None if x is None else str(x), doc=None):
-        self.method = target_method
-        self.pattern = re.compile(b(regex), regex_flags)
+    def __init__(self, func, pattern, argument_mappings=None, return_mapping=None, doc=None):
+        if not callable(func):
+            raise RuntimeError('Can not construct a Func-object from a non callable object.')
+
+        self.func = func
+        self.raw_pattern = pattern
+        self.pattern = re.compile(b(pattern), 0) if pattern else None
+
+        func_args = num_args(func)
+        if self.pattern.groups != func_args:
```

Comment:
You might be better off using `inspect.getcallargs` ([link](https://docs.python.org/2/library/inspect.html#inspect.getcallargs)) for this purpose. That function will allow you to *"try"* calling `func` without actually `try`ing to call it, if that makes sense.

Consider:
```python
inspect.getcallargs(func, *[None]*self.pattern.groups)
```
This will return some stuff (that you don't care about here) *IF it is possible* to call `func` with that number of arguments (without actually calling it). If it is NOT possible to call it with that amount of arguments, it will raise a `TypeError` exception.

This way, you don't need `num_args` at all from what I can tell (and the tests for it can go away as well). A shortcoming of this `num_args` implementation is that it doesn't consider optional arguments with default values, or `*args` / `**kwargs` scenarios... if `func` has anything like that, it will reject valid regexes / group counts.